fastssl/models/resnets.py

```diff
@@ -0,0 +1,104 @@
+                nn.Identity() if no_bn else nn.BatchNorm2d(self.expansion*planes)
+            )
+
+    def forward(self, x):
+        out = F.relu(self.bn1(self.conv1(x)))
+        out = self.bn2(self.conv2(out))
+        out += self.shortcut(x)
+        out = F.relu(out)
+        return out
+
+
+class Bottleneck(nn.Module):
+    expansion = 4
+
+    def __init__(self, in_planes, planes, stride=1, no_bn=False):
+        super(Bottleneck, self).__init__()
+        self.conv1 = nn.Conv2d(in_planes, planes, kernel_size=1, bias=False)
+        self.bn1 = nn.Identity() if no_bn else nn.BatchNorm2d(planes)
+        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3,
+                               stride=stride, padding=1, bias=False)
+        self.bn2 = nn.Identity() if no_bn else nn.BatchNorm2d(planes)
+        self.conv3 = nn.Conv2d(planes, self.expansion *
+                               planes, kernel_size=1, bias=False)
+        self.bn3 = nn.Identity() if no_bn else nn.BatchNorm2d(self.expansion*planes)
+
+        self.shortcut = nn.Sequential()
+        if stride != 1 or in_planes != self.expansion*planes:
+            self.shortcut = nn.Sequential(
+                nn.Conv2d(in_planes, self.expansion*planes,
+                          kernel_size=1, stride=stride, bias=False),
+                nn.Identity() if no_bn else nn.BatchNorm2d(self.expansion*planes)
+            )
+
+    def forward(self, x):
+        out = F.relu(self.bn1(self.conv1(x)))
+        out = F.relu(self.bn2(self.conv2(out)))
+        out = self.bn3(self.conv3(out))
+        out += self.shortcut(x)
+        out = F.relu(out)
+        return out
+
+
+class ResNet(nn.Module):
+    def __init__(self, block, num_blocks, width=64, num_classes=10, no_bn=False):
+        super(ResNet, self).__init__()
+        self.in_planes = width
+        self.no_bn = no_bn
+
+        self.conv1 = nn.Conv2d(3, width, kernel_size=3,
+                               stride=1, padding=1, bias=False)
+        self.bn1 = nn.Identity() if no_bn else nn.BatchNorm2d(width)
+        self.layer1 = self._make_layer(block, width, num_blocks[0], stride=1)
+        self.layer2 = self._make_layer(block, 2*width, num_blocks[1], stride=2)
+        self.layer3 = self._make_layer(block, 4*width, num_blocks[2], stride=2)
+        self.layer4 = self._make_layer(block, 8*width, num_blocks[3], stride=2)
+        self.avgpool = nn.AdaptiveAvgPool2d(output_size=(1, 1))
+        self.linear = nn.Linear(8*width*block.expansion, num_classes)
+
+    def _make_layer(self, block, planes, num_blocks, stride):
+        strides = [stride] + [1]*(num_blocks-1)
+        layers = []
+        for stride in strides:
+            layers.append(block(self.in_planes, planes, stride, no_bn=self.no_bn))
+            self.in_planes = planes * block.expansion
+        return nn.Sequential(*layers)
+
+    def forward(self, x):
+        out = F.relu(self.bn1(self.conv1(x)))
```

Comment:
Unless we want to log the intermediate features, I'd suggest putting all layers in nn.Sequential.

Reply:
All layers of Resnet? I thought that maybe we could keep them separate in case we want to investigate individual blocks in the future. Having them separate might come in handy to attach forward hooks.